recipes-ni/ni-base-system-image-tests/files/fs_permissions_known.py

```diff
@@ -0,0 +1,153 @@
+import argparse
+import csv
+import datetime
+import hashlib
+import mmap
+import os
+import re
+import subprocess
+import sys
+
+class Logger:
+    prefix_logs = []
+    logs = []
+    def log(self, log):
+        self.logs.append(log)
+
+    def prefix_log(self, log):
+        self.prefix_logs.append(log)
+
+    def report(self):
+        for log in self.prefix_logs:
+            print(log)
+        for log in self.logs:
+            print(log)
+
+def run_cmd(cmd):
+    output = subprocess.check_output(cmd)
+    return output.decode('utf-8')
+
+# Columns: Path, Mode, User, Group, Link Target
+fs_manifest_format = '%p\t%M\t%u\t%g\t%l'
+def get_fs_manifest():
+    search_dirs = ['/bin', '/boot', '/etc', '/lib', '/lib64', '/sbin', '/usr', '/var']
+    omit_dirs = ['/etc/natinst/niskyline/Data/Assets/Cache', '/lib/modules', '/var/cache', '/var/run', '/var/tmp', '/var/volatile']
```

Comment:
Well, we have to check the permissions somehow, idk if there's a better way to do that than with `find`, but we could call `find` only on the files we want instead of batching the find and looping through them to organize it into a dict.

I'll put the logic to check a path's permissions into a method, so if it's not performant, we can adjust it easily:

```
class Permissions:
    def __init__(self, file_name):
        raw_entry = get_permissions(file_name)
        entry = raw_entry.split('\t')

        if len(entry) > 4:
            self.path = entry[0]
            self.mode = entry[1]
            self.user = entry[2]
            self.group = entry[3]
            self.link = entry[4]
        else:
            self.path = None
            self.mode = None
            self.user = None
            self.group = None
            self.link = None

    def get_permissions(file_name):
        find_format = '%p\t%M\t%u\t%g\t%l\n'
        find_cmd = [ 'find', file_name, '-printf', find_format, '-prune' ]
        return run_cmd(find_cmd).strip()

    def run_cmd(cmd):
        output = subprocess.check_output(cmd)
        return output.decode('utf-8')
```

Reply:
A bit of searching shows a couple of relevant Python functions here: [`os.stat`](https://docs.python.org/3/library/os.html?highlight=os%20stat#os.stat) and possibly [`pwd.getpwuid`](https://docs.python.org/3/library/pwd.html#pwd.getpwuid).  
I think that if we are going to expand the known paths ahead of time, and give just a single path to this class, we might as well do that stat call ourselves rather than parsing command output.

Reply:
Okay, here's a python way to do it:

```
class Permissions:
    def __init__(self, file_name):
        if !os.path.exists(file_name):
            self.path = None
            self.mode = None
            self.user = None
            self.group = None
            self.link = None
        stats = os.stat(file_name)
        self.path = file_name
        self.mode = stat.filemode(stats.st_mode)
        self.user = pwd.getpwuid(stats.st_uid)
        self.group = pwd.getpwgid(stats.st_gid)
        if stat.S_ISLINK(self.mode):
            self.link = os.readlink(self.path)
        else:
            self.link = ''
```